Refactor ternary operation

Aswinmcw opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/31028

### Problem description
currently only where op have the ternary infra

### What's changed
Move the where op as common ternary infra to support other ternary ops in future

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
Refactor ternary operations by replacing 'where' includes with 'ternary' in multiple files and removing the obsolete 'where' implementation.

### What's changed
- Updated includes in `binary_composite_op.cpp`, `binary_backward.cpp`, `complex_unary_backward_op.cpp`, and other files to use the new `ternary.hpp` instead of the deprecated `where.hpp`.
- Introduced new `ternary` implementation files and removed the old `where` implementation to streamline the ternary operation handling.

### Checklist
- [x] All post-commit CI passes
- [x] New/Existing tests provide coverage for changes

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/ternary_revamp

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/ternary_revamp

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/ternary_revamp

Aswinmcw committed:
Fix TST variant validation message in ternary_device_operation.cpp to correctly reference input_b as scalar

Aswinmcw commented:
@tt-ai review this PR

Aswinmcw committed:
Refactor ternary operation types to use ScalarVariant for true and false values

### What's changed
- Updated function signatures in `ternary_pybind.cpp`, `ternary.cpp`, and `ternary.hpp` to replace `std::variant<float, Tensor>` with `ScalarVariant`.
- Introduced `ScalarVariant` type alias in `ternary_op_types.hpp` to support float, int, and Tensor types.

This change enhances type clarity and consistency across ternary operations.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/ternary_revamp

Aswinmcw committed:
Refactor ternary operation types to use TensorScalarVariant

Updated the ternary operation implementation to replace ScalarVariant with TensorScalarVariant for true and false values across multiple files. This change enhances type safety and consistency in handling tensor and scalar values. Additionally, updated related compute definitions to reflect the new naming convention for ternary operations.

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/device/kernels/compute/ternary_sfpu_col_scalar_bcast_ttt.cpp:
```diff
@@ -58,9 +58,9 @@ ALWI void process_tile(
-        WHERE_LLK(0, 1, 2, 0);  // where(predicate, true, false)
+        // Perform the ternary operation
+        TERNARY_SFPU_OP_INIT();
+        TERNARY_SFPU_OP_FUNC(0, 1, 2, 0);  // ternary_op(predicate, true, false)
```
can be removed or modified: (predicate, true, false)

Aswinmcw commented:
/codeowners

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/ternary/common/ternary_op_types.hpp:
```diff
@@ -0,0 +1,41 @@
+    TTS,  // tensor-tensor-scalar
+    TST,  // tensor-scalar-tensor
+    TSS,  // tensor-scalar-scalar
+};
```
While uninteresting for the specific case of `WHERE`, `STT`, `STS`, and `SST` would be useful to have for operations like `LERP`, which is another reason why I'm suggesting we should consider delegating these cases to the other primitive operations instead.

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/ternary/common/ternary_op_types.hpp:
```diff
@@ -0,0 +1,41 @@
+    TTT,  // tensor-tensor-tensor
+    TTS,  // tensor-tensor-scalar
+    TST,  // tensor-scalar-tensor
+    TSS,  // tensor-scalar-scalar
```
The TSS variants will be using Unary infra. 
We did think about using Binary infra for TTS and TST variants.. We'll need to modify binary infra a bit - binary infra only allows TT or TS cases right now.

KalaivaniMCW replied:
Yes @patrickroberts 
"This suggestion can also be applied to the binary infrastructure: Rather than implement a primitive "binary" with up to one optional input tensor, can we delegate the TS and ST cases there to unary primitive operation instead" 

I was trying to this out for Relational ops which is where I noticed that the scalar values being `float` cause incorrect values being sent to the LLK when I tried out INT32/UINT32  https://github.com/tenstorrent/tt-metal/issues/23179#issuecomment-3433618957
I'll push that case to a branch and share

dchenTT replied:
I am going to support ST in binary_ng soon.  scalar_value + tensor.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/ternary_revamp

Aswinmcw committed:
Add broadcasting functions for ternary and binary operations

### What's changed
- Introduced `compute_broadcasted_output_ternary` and `compute_broadcasted_output_binary` functions to handle shape broadcasting for ternary and binary operations, respectively.
- Removed inline lambda implementations for broadcasting in favor of these new static functions for better readability and maintainability.
- Added error handling for broadcasting rule violations with detailed error messages.

### Impact
This change enhances the robustness of shape handling in ternary operations, ensuring compliance with broadcasting rules across different tensor shapes.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/ternary_revamp

Aswinmcw committed:
Refactor benchmark and logging in ternary operations

- Updated the include path in benchmark_where.cpp to reference the correct ternary operations header.
- Changed logging level from info to debug in ternary.cpp for various log statements to reduce verbosity during execution.

Aswinmcw committed:
Update CMakeLists.txt to include additional header file for ternary operations

- Added 'common/ternary_op_types.hpp' to the list of header files in the target_sources for ternary operations, ensuring proper inclusion of necessary types.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/ternary_revamp

Aswinmcw committed:
Added include directive for tensor header in ternary operation types header file to ensure proper type definitions are available for ternary operations.
